pep-9999.rst

```diff
@@ -0,0 +1,669 @@
+uses of ``Py_CHECKWRITE`` in the CPython codebase. Two good
+starting places are ``object.c`` `[1]`_ and ``dictobject.c`` `[2]`_.
+
+.. _[1]: https://github.com/mjp41/cpython/pull/51/files#diff-ba56d44ce0dd731d979970b966fde9d8dd15d12a82f727a052a8ad48d4a49363
+.. _[2]: https://github.com/mjp41/cpython/pull/51/files#diff-b08a47ddc5bc20b2e99ac2e5aa199ca24a56b994e7bc64e918513356088c20ae
+
+Deep Freezing Semantics
+-----------------------
+
+The ``freeze(obj)`` builtin works as follows:
+
+1. It recursively marks ``obj`` and all objects reachable from ``obj``
+   immutable.
+2. If ``obj`` is already immutable (e.g., an integer, string, or a
+   previously frozen object), the recursion terminates.
+3. The freeze operation follows object references (relying on ``tp_traverse`` 
+   in the type structs of the objects involved), including:
+
+    * Object attributes (``__dict__`` for user-defined objects,
+      ``tp_dict`` for built-in types).
+    * Container elements (e.g., lists, tuples, dictionaries,
+      sets).
+    * The ``__class__`` attribute of an object (which makes freezing
+      instances of user-defined classes also freeze their class
+      and its attributes).
+    * The ``__bases__`` chain in classes (freezing a class freezes its
+      base classes).
+
+5. Attempting to mutate a frozen object raises an exception (``NotWriteableError``).
+
+
+Illustration of the Deep Freezing Semantics
+-------------------------------------------
+
+Consider the following code:
+
+.. code-block:: python
+
+	class Foo: pass
+	x = Foo()
+	x.f = 42
+
+If we glance over some implementation details such as the classes of strings and ints being
+implemented in C, and the naming of fields (the ``__metaclass__`` fields are named ``__class__`` etc.), 
+the code above gives rise to the following object graph:
+
+.. image:: pep-9999/freeze1.svg
+   :width: 66%
+
+The ``Foo`` instance pointed to by ``x`` consists of several objects: its fields are stored in a 
+dictionary object, and the assignment ``x.f = 42`` adds two objects to the dictionary in the form
+of a string key ``"f"`` and its associated value ``42``. These objects each have a class pointers
+pointing to the ``string`` and ``int`` classes respectively. Similarly, the ``foo`` instance has
+a pointer to the ``Foo`` class. Finally, all these classes have a pointer to the same meta class
+object.
+
+Calling ``freeze(x)`` will freeze **all** of these objects.
+
+(Note that in this particular case, both ``"f"`` and ``42`` will be frozen already, see
+default immutability below.)
+
+
+Default (Im)Mutabiliy 
+---------------------
+
+Interned strings, numbers in the small integer cache, and tuples of
+immutable objects are made immutable in this PEP. This is either
+consistent with current Python semantics or backwards-compatible.
+
+A reasonable design would make all numbers immutable, not just those
+in the small integer cache. This should be properly investigated.
```

Comment:
Aren't all numbers already immutable?  (They aren't all immortal, but I think they are immutable -- or are you referring to optimizations where the memory is re-used and mutated?)